File: app/src/main/java/vergecurrency/vergewallet/wallet/AddressValidator.java
package vergecurrency.vergewallet.wallet;


import android.provider.Telephony;

import java.util.HashMap;

import io.horizontalsystems.bitcoinkit.exceptions.AddressFormatException;
import io.horizontalsystems.bitcoinkit.network.MainNet;
import io.horizontalsystems.bitcoinkit.utils.AddressConverter;

public class AddressValidator {

	public static boolean isValidAddress(String address) {
		try {
			new AddressConverter(new MainNet()).convert(address);
			return true;
		} catch (AddressFormatException e) {
			System.err.println(e.getMessage());
			return false;
		}
	}

	public ValidationCompletion validate(String address) {
		ValidationCompletion vc = new ValidationCompletion(false, "", 0f);
		if (AddressValidator.isValidAddress(address)) {
			vc.setValid(true);
			vc.setAddress(address);
		}

		String[] splittedRequest = address.replace("verge://", "").replace("verge:", "").split("\\?");

		String [] parameters = splittedRequest[splittedRequest.length-1].split("&");

		if (AddressValidator.isValidAddress(splittedRequest[0])) {
			vc.setValid(true);
			vc.setAddress(splittedRequest[0]);
		} else {
			return vc;
		}

		String [] splittedParameters = parameters[parameters.length-1].split("&");

		HashMap<String,String> definitiveParameters = new HashMap<>();

		for (String param : splittedParameters) {
			String[] parameterPair = param.split("=");
			definitiveParameters.put(parameterPair[0], parameterPair[1]);
		}

		Float amount = Float.parseFloat(definitiveParameters.getOrDefault("amount","0"));

		vc.setAmount(amount);

		return vc;
	}


	//inner struct
	public class ValidationCompletion {
		boolean valid;
		String address;
		Float amount;

		public ValidationCompletion(boolean valid, String address, Float amount) {
			this.valid = valid;
			this.address = address;
			this.amount = amount;
		}

		public boolean isValid() {
			return valid;
		}

		public void setValid(boolean valid) {
			this.valid = valid;
		}

		public String getAddress() {
			return address;
		}

		public void setAddress(String address) {
			this.address = address;
		}

		public Float getAmount() {
			return amount;
		}

		public void setAmount(Float amount) {
			this.amount = amount;
		}
	}
}

File: app/src/main/kotlin/vergecurrency/vergewallet/wallet/AddressValidator.kt
package vergecurrency.vergewallet.wallet


import io.horizontalsystems.bitcoinkit.network.MainNet
import io.horizontalsystems.bitcoinkit.utils.AddressConverter
import java.lang.Float.parseFloat
import java.util.*

typealias ValidationCompletion = (valid: Boolean, address: String?, amount: Float?) -> Unit

class AddressValidator {

    //implement fun override for barcode object

    fun validate(string: String, completion: ValidationCompletion) {
        var valid = false
        var address: String? = null
        var amount: Float? = null


        if (isValidAddress(string)) {
            valid = true
            address = string
        }

        val splitRequest = string.replace("verge://", "").replace("verge:", "").split("\\?".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        val parametersString = splitRequest.last().split("&".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        if (isValidAddress(splitRequest.first())) {
            valid = true
            address = splitRequest.first()
        } else {
            return completion(valid, address, amount)
        }

        val splitParameters = parametersString.last().split("&".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        val parameters = HashMap<String, String>()

        for (param in splitParameters) {
            val parameterPair = param.split("=".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            parameters[parameterPair.first()] = parameterPair.last()
        }

        val amountParam = parameters["amount"]
        if (amountParam != null) {
            amount = parseFloat(amountParam)
        }

        return completion(valid, address, amount)

    }

    private fun isValidAddress(address: String): Boolean {
        return try {
            AddressConverter(MainNet()).convert(address)
            true
        } catch (e: Exception) {
            false
        }
    }

}
